src/pymor/playground/operators/block.py

```diff
@@ -0,0 +1,149 @@
+                range_types[i] = op.range
+
+        self.source = VectorSpace(BlockVectorArray, tuple(source_types))
+        self.range = VectorSpace(BlockVectorArray, tuple(range_types))
+        self._source_dims = tuple(space.dim for space in self.source.subtype)
+        self._range_dims = tuple(space.dim for space in self.range.subtype)
+        self.num_source_blocks = len(source_types)
+        self.num_range_blocks = len(range_types)
+        self.linear = all(op.linear for op in self._operators())
+        self._is_diagonal = (all(block is None if i != j else True for (i, j), block in np.ndenumerate(self._blocks))
+                             and self.num_source_blocks == self.num_range_blocks)
+        self.build_parameter_type(inherits=list(self._operators()))
+
+    @classmethod
+    def hstack(cls, operators):
+        """Horizontal stacking of operators
+
+        Parameters
+        ----------
+        operators
+            A tuple, list, array, or iterable of operators.
+        """
+        blocks = np.array([[op for op in operators]])
+        return cls(blocks)
+
+    @classmethod
+    def vstack(cls, operators):
+        """Vertical stacking of operators
+
+        Parameters
+        ----------
+        operators
+            A tuple, list, array, or iterable of operators.
+        """
+        blocks = np.array([[op] for op in operators])
+        return cls(blocks)
+
+    def apply(self, U, ind=None, mu=None):
+        assert U in self.source
+        assert U.check_ind(ind)
+
+        blocks = []
+        for i in xrange(self.num_range_blocks):
+            block = None
+            for j in xrange(self.num_source_blocks):
+                op = self._blocks[i, j]
+                if op is not None:
+                    V = op.apply(U.block(j), ind=ind, mu=mu)
+                    if block is None:
+                        block = V.copy()
+                    else:
+                        block += V
+            blocks.append(block.copy())
+
+        return BlockVectorArray(blocks)
+
+    def apply_adjoint(self, U, ind=None, mu=None, source_product=None, range_product=None):
+        assert U in self.range
+        assert U.check_ind(ind)
+        assert source_product is None or source_product.source == source_product.range == self.source
+        assert range_product is None or range_product.source == range_product.range == self.range
+
+        if range_product is not None:
+            U = range_product.apply(U)
+
+        blocks = []
+        for j in xrange(self.num_source_blocks):
+            block = None
+            for i in xrange(self.num_range_blocks):
+                op = self._blocks[i, j]
+                if op is not None:
+                    V = op.apply_adjoint(U.block(i), ind=ind, mu=mu)
+                    if block is None:
+                        block = V.copy()
+                    else:
+                        block += V
+            blocks.append(block.copy())
```

Comment:
Ah, sorry for that. I meant s.a. as in see above. As in: 
* if we do not need `copy()` above, then we do not need it here, either
* if you agree that a pre-initialization would make sense above, then I would also pre-initialize `blocks` as a list of appropriate length (`num_source_blocks`) here. The same holds further below...